agent/consul/intention_endpoint.go

```diff
@@ -171,336 +317,181 @@ func (s *Intention) prepareApplyUpdateLegacy(
-			noop = true
-			if prevEntry != nil {
-				sn := args.Intention.SourceServiceName()
-
-				upsertEntry = prevEntry.Clone()
-				for i, src := range upsertEntry.Sources {
-					if src.SourceServiceName() == sn {
-						// Delete slice element: https://github.com/golang/go/wiki/SliceTricks#delete
-						//    a = append(a[:i], a[i+1:]...)
-						upsertEntry.Sources = append(upsertEntry.Sources[:i], upsertEntry.Sources[i+1:]...)
-
-						if len(upsertEntry.Sources) == 0 {
-							upsertEntry.Sources = nil
-						}
-						noop = false
-						break
-					}
-				}
-			}
+	deleted := upsertEntry.DeleteSourceByLegacyID(args.Intention.ID)
+	if !deleted {
+		return "", nil, fmt.Errorf("Cannot delete non-existent intention: '%s'", args.Intention.ID)
+	}
 
-		} else {
-			// legacy style: LegacyID required
-			legacyWrite = true
+	if upsertEntry == nil || len(upsertEntry.Sources) == 0 {
+		return structs.ConfigEntryDelete, &structs.ServiceIntentionsConfigEntry{
+			Kind:           structs.ServiceIntentions,
+			Name:           prevEntry.Name,
+			EnterpriseMeta: prevEntry.EnterpriseMeta,
+		}, nil
+	}
 
-			prevEntry, err = s.prepareApplyDeleteLegacy(ident, authz, args)
-			if err != nil {
-				return err
-			}
+	return structs.ConfigEntryUpsertCAS, upsertEntry, nil
+}
 
-			upsertEntry = prevEntry.Clone()
-			for i, src := range upsertEntry.Sources {
-				if src.LegacyID == args.Intention.ID {
-					// Delete slice element: https://github.com/golang/go/wiki/SliceTricks#delete
-					//    a = append(a[:i], a[i+1:]...)
-					upsertEntry.Sources = append(upsertEntry.Sources[:i], upsertEntry.Sources[i+1:]...)
-
-					if len(upsertEntry.Sources) == 0 {
-						upsertEntry.Sources = nil
-					}
-					break
-				}
-			}
-		}
+func (s *Intention) computeApplyChangesDelete(
+	entMeta *structs.EnterpriseMeta,
+	args *structs.IntentionRequest,
+) (structs.ConfigEntryOp, *structs.ServiceIntentionsConfigEntry, error) {
+	args.Intention.DefaultNamespaces(entMeta)
 
-	case structs.IntentionOpDeleteAll:
-		// This is an internal operation initiated by the leader and is not
-		// exposed for general RPC use.
-		fallthrough
-	default:
-		return fmt.Errorf("Invalid Intention operation: %v", args.Op)
+	prevEntry, err := s.getServiceIntentionsConfigEntry(args.Intention.DestinationName, args.Intention.DestinationEnterpriseMeta())
+	if err != nil {
+		return "", nil, err
 	}
 
-	if !noop && prevEntry != nil && legacyWrite && !prevEntry.LegacyIDFieldsAreAllSet() {
-		sn := prevEntry.DestinationServiceName()
-		return fmt.Errorf("cannot use legacy intention API to edit intentions with a destination of %q after editing them via a service-intentions config entry", sn.String())
+	if prevEntry == nil {
+		return "", nil, nil // no op means no-op
 	}
 
-	// setup the reply which will have been filled in by one of the preparedApply* funcs
-	if legacyWrite {
-		*reply = args.Intention.ID
-	} else {
-		*reply = ""
-	}
+	// NOTE: validation errors may be misleading!
```

Comment:
Where is the validation invoked? Maybe this comment should be moved closer to where it would apply

Reply:
So I left this in for myself and didn't come back and figure out how to improve things. In the #9151 followup I think I actually removed these comments for now.

The gist is that all of the real validation happens in the config entry Validate() method, and the error messages refer to fields on the config entry, not the `structs.Intention` object.